maskit/optimizers.py

```diff
@@ -1,5 +1,165 @@
+    :param maxiter: Maximum number of iterations, defaults to 15000
+    :param disp: [description], defaults to None
+    :param callback: Called after each iteration with current parameters,
+        defaults to None
+    :param maxls: Maximum number of line search steps (per iteration),
+        defaults to 20
+    """
+
+    __slots__ = (
+        "bounds",
+        "m",
+        "factr",
+        "pgtol",
+        "epsilon",
+        "iprint",
+        "maxfun",
+        "maxiter",
+        "disp",
+        "callback",
+        "maxls",
+    )
+
+    def __init__(
+        self,
+        bounds: Optional[ndarray] = None,
+        m: int = 10,
+        factr: float = 1e7,  # qiskit: factr: float = 10,
+        pgtol: float = 1e-5,
+        epsilon: float = 1e-8,
+        iprint: int = -1,
+        maxfun: int = 15000,  # qiskit: maxfun: int = 1000,
+        maxiter: int = 15000,
+        disp=None,
+        callback=None,
+        maxls: int = 20,
+    ):
+        self.bounds = bounds
+        self.m = m
+        self.factr = factr
+        self.pgtol = pgtol
+        self.epsilon = epsilon
+        self.iprint = iprint
+        self.maxfun = maxfun
+        self.maxiter = maxiter
+        self.disp = disp
+        self.callback = callback
+        self.maxls = maxls
+
+    def optimize(
+        self,
+        objective_fn,
+        parameters: ndarray,
+        *args,
+        grad_fn=None,
+        **kwargs,
+    ) -> Tuple[ndarray, float, ndarray]:
+        """
+        :param objective_fn: Function to minimize
+        :param parameters: Initial guess of parameters
+        :param grad_fn: The gradient of `func`. In case of `None`, the gradient is
+            approximated numerically, defaults to None
+        """
+        return self._optimise(
+            objective_fn, parameters, self.maxiter, *args, grad_fn=grad_fn, **kwargs
+        )
+
+    def _optimise(
+        self, objective_fn, parameters: ndarray, maxiter, *args, grad_fn, **kwargs
+    ) -> Tuple[ndarray, float, ndarray]:
+        shape = parameters.shape
+        shaped_fn = self._wrap_objective_fn(objective_fn, shape, **kwargs)
+        approx_grad = False if grad_fn is not None else True
+        updated_parameters, cost, info = sciopt.fmin_l_bfgs_b(
+            shaped_fn,
+            parameters.flatten(),
+            fprime=get_gradient(shaped_fn) if grad_fn is None else grad_fn,
+            *args,
+            approx_grad=approx_grad,
+            bounds=self.bounds,
+            m=self.m,
+            factr=self.factr,
+            pgtol=self.pgtol,
+            epsilon=self.epsilon,
+            iprint=self.iprint,
+            maxfun=self.maxfun,
+            maxiter=maxiter,
+            disp=self.disp,
+            callback=self.callback,
+            maxls=self.maxls,
+        )
+        return updated_parameters.reshape(shape), cost, info["grad"]
+
+    def _wrap_objective_fn(self, objective_fn, shape, **kwargs):
```

Comment:
How about a different name? "wrap" seems somewhat nondescript.
```suggestion
    def _reshaping_objective_fn(self, objective_fn, shape, **kwargs):
```